src/uct/ib/rdmacm/rdmacm_cm_ep.c

```diff
@@ -7,13 +7,212 @@
 #include "rdmacm_cm_ep.h"
 
 
-UCS_CLASS_INIT_FUNC(uct_rdmacm_cm_ep_t, const uct_ep_params_t *params)
+static ucs_status_t uct_rdmacm_create_dummy_cq_qp(struct rdma_cm_id *id,
+                                                  struct ibv_cq **cq_p,
+                                                  struct ibv_qp **qp_p)
+{
+    struct ibv_qp_init_attr qp_init_attr;
+    ucs_status_t status;
+    struct ibv_cq *cq;
+    struct ibv_qp *qp;
+
+    /* Create a dummy completion queue */
+    cq = ibv_create_cq(id->verbs, 1, NULL, NULL, 0);
+    if (cq == NULL) {
+        ucs_error("ibv_create_cq() failed: %m");
+        status =  UCS_ERR_IO_ERROR;
+        goto out;
+    }
+
+    /* Create a dummy UD qp */
+    memset(&qp_init_attr, 0, sizeof(qp_init_attr));
+    qp_init_attr.send_cq = cq;
+    qp_init_attr.recv_cq = cq;
+    qp_init_attr.qp_type = IBV_QPT_UD;
+    qp_init_attr.cap.max_send_wr  = 2;
+    qp_init_attr.cap.max_recv_wr  = 2;
+    qp_init_attr.cap.max_send_sge = 1;
+    qp_init_attr.cap.max_recv_sge = 1;
+
+    qp = ibv_create_qp(id->pd, &qp_init_attr);
+    if (qp == NULL) {
+        ucs_error("failed to create a dummy ud qp. %m");
+        status = UCS_ERR_IO_ERROR;
+        goto out_destroy_cq;
+    }
+
+    ucs_debug("created ud QP %p with qp_num: 0x%x and cq %p on rdmacm_id %p",
+              qp, qp->qp_num, cq, id);
+
+    *cq_p = cq;
+    *qp_p = qp;
+
+    return UCS_OK;
+
+out_destroy_cq:
+    ibv_destroy_cq(cq);
+out:
+    return status;
+}
+
+ucs_status_t
+uct_rdamcm_cm_ep_set_remaining_conn_param(struct rdma_conn_param *conn_param,
+                                           const uct_rdmacm_priv_data_hdr_t *hdr,
+                                           uct_rdmacm_cm_ep_t *cep)
+{
+    ucs_status_t status;
+    struct ibv_qp *qp;
+    struct ibv_cq *cq;
+
+    /* create a dummy qp in order to get a unique qp_num to provide to librdmacm */
+    status = uct_rdmacm_create_dummy_cq_qp(cep->id, &cq, &qp);
```

Comment:
better to call this like
`uct_rdmacm_create_dummy_cq_qp(self->id->pd, &self->cq, &self->qp);` from the EP constructor

Reply:
I can't create the cq and qp in ep init because i'm using the rdmacm_id for the ibv_context and it's only present after receiving the RDMA_CM_EVENT_ADDR_RESOLVED event.

Reply:
I see, thanks for clarification